[promtail] Add more customizations to values.yaml

piontec opened the pull request:
This PR adds the following options, which were available previously:
- add custom keys to the `client` section of the config file
- add extra scrape configs directly (to allow for journal or static scrapping)

unguiculus commented on charts/promtail/values.yaml:
```diff
@@ -214,6 +214,14 @@ podSecurityPolicy:
   requiredDropCapabilities:
     - ALL
 
+# -- You can put here any keys that will be directly added to the config file's 'client' block.
+# @default -- empty
+extra_client_configs: {}
+
+# -- You can put here any additional scrape configs you want to add to the config file.
+# @default -- empty
+extra_scrape_configs: {}
```
If we add something like this, then it must go under `config.snippets`. Also: camel case please.

piontec replied:
RIght, but this requires overriding the whole `config.file`. I really liked the previous approach, where you could override only some sections and keep your actual config shorter. As for camel case - sure, I'm not sure how I missed that :P

unguiculus replied:
OK, I guess it makes sense.

unguiculus reviewed (changes requested):
Something like this is already possible. Have a look at the values file for tests:
https://github.com/grafana/helm-charts/blob/main/charts/promtail/ci/service-values.yaml

piontec commented:
Ping @unguiculus , can you please have a look again?

unguiculus reviewed (changes requested):
It doesn't look like you considered my comments. Please use camel case and bump the minor version. Please don't rebase or force-push. Just add additional commits so GitHub can properly display your changes. Thanks.

unguiculus commented on charts/promtail/values.yaml:
```diff
@@ -431,9 +439,11 @@ config:
 
     client:
       url: {{ .Values.config.lokiAddress }}
+      {{- toYaml .Values.extraClientConfigs | nindent 2 }}
 
     positions:
       filename: /run/promtail/positions.yaml
 
     scrape_configs:
       {{- tpl .Values.config.snippets.scrapeConfigs $ | nindent 2 }}
+      {{- toYaml .Values.extraScrapeConfigs | nindent 2 }}
```
As suggested and to be consistent, these should go under `config.snippets`. We may also want to make them templated. As such, they should be strings.

```suggestion
      {{- tpl .Values.config.snippets.extraClientConfigs . | nindent 2 }}

    positions:
      filename: /run/promtail/positions.yaml

    scrape_configs:
      {{- tpl .Values.config.snippets.scrapeConfigs . | nindent 2 }}
      {{- tpl .Values.config.snippets.extraScrapeConfigs . | nindent 2 }}
```

unguiculus reviewed (changes requested):
Would you mind updating the examples in the readme to use the new configs?

unguiculus commented on charts/promtail/Chart.yaml:
```diff
@@ -3,7 +3,7 @@ name: promtail
 type: application
 appVersion: 2.1.0
-version: 3.0.4
+version: 3.0.5
```
Minor bump, please. It's a new feature.

piontec replied:
Good idea, thanks!

unguiculus commented on charts/promtail/values.yaml:
```diff
@@ -214,6 +214,14 @@ podSecurityPolicy:
   requiredDropCapabilities:
     - ALL
 
+# -- You can put here any keys that will be directly added to the config file's 'client' block.
+# @default -- empty
+extraClientConfigs: {}
+
+# -- You can put here any additional scrape configs you want to add to the config file.
+# @default -- empty
+extraScrapeConfigs: {}
```
You will have to move these to the right place as well.

piontec commented:
> Would you mind updating the examples in the readme to use the new configs?

I simplified the existing syslog and journald examples with `extraScrapeConfigs` and added basic auth + tenant ID for `extraClientConfig`.

unguiculus commented:
One more thing: Please update `charts/promtail/ci/service-values.yaml` as well.
